Adding section Simulating a Man-in-the-Middle Attack with an Access P… (#1599)
OWASP/owasp-mstg: Document/0x04f-Testing-Network-Communication.md

- Route the traffic through the host machine. You can set up your machine as the network gateway, e.g. by using the built-in Internet Sharing facilities of your operating system. You can then use [Wireshark](https://www.wireshark.org "Wireshark") to sniff any traffic from the mobile device;

> - Sometimes you need to execute a MITM attack to force the mobile device to talk to you. For this scenario you should consider [bettercap](https://github.com/bettercap/bettercap "bettercap") or use your own Access Point to redirect network traffic from the mobile device to your host machine (see below);

Reviewer: Could you please correct "Access Point" to lowercase for all occurrences in this section? Thanks!
```suggestion
- Sometimes you need to execute a MITM attack to force the mobile device to talk to you. For this scenario you should consider [bettercap](https://github.com/bettercap/bettercap "bettercap") or use your own access point to redirect network traffic from the mobile device to your host machine (see below);
```

Author: New commit. Comment implemented.

> bettercap is a powerful tool to execute MITM attacks and should be preferred nowadays, instead of ettercap. See also [Why another MITM tool?](https://www.bettercap.org/legacy/#why-another-mitm-tool "Why another MITM tool?") on the bettercap site.

- On a rooted device, you can use hooking or code injection to intercept network-related API calls (e.g. HTTP requests) and dump or even manipulate the arguments of these calls. This eliminates the need to inspect the actual network data. We'll talk in more detail about these techniques in the "Reverse Engineering and Tampering" chapters;

- On macOS, you can create a "Remote Virtual Interface" for sniffing all traffic on an iOS device. We'll describe this method in the chapter "Basic Security Testing on iOS".

#### Simulating a Man-in-the-Middle Attack with bettercap

##### Network Setup

> To be able to get a man-in-the-middle position your machine should be in the same wireless network as the mobile phone and the gateway it communicates to. Once this is done you need the IP address of mobile phone.

Reviewer 2: ```suggestion
To be able to get a man-in-the-middle position your machine should be in the same wireless network as the mobile phone and the gateway it communicates to. Once this is done you need the IP address of your mobile phone.
```

Following scenarios are possible:

- use your machine built-in WiFi card as an access point and use your wired connection to connect to the target network,
- Use an external USB WiFi card as an access point and user your machine built-in WiFi to connect to the target network (can be vice-versa).
- Use a separate access point and redirect the traffic to your machine.

The scenario with an external USB card require that the WiFi card has the capability to create an access point. Additionally, you need to install some tools and/or configure the network to enforce a man-in-the-middle position (see below). You can verify if your WiFi card has AP capabilities by using the command `iwconfig` on Kali Linux:

    ```shell
    $ iw list | grep AP 
    ```

The scenario with a separate access point requires access to the configuration of the AP and you should check first if the AP supports either:

- port forwarding or
- has a span or mirror port.

> In both cases the AP needs to be configured to point to your machines IP. Your machine must be connected to the AP (via wired connection or WiFi) and you need to have connection to the target network (can be by the same connection as to the AP). Some additional configuration may be required on your machine to route traffic to the target network.

Reviewer 2: ```suggestion
In both cases the AP needs to be configured to point to your machine's IP. Your machine must be connected to the AP (via wired connection or WiFi) and you need to have connection to the target network (can be by the same connection as to the AP). Some additional configuration may be required on your machine to route traffic to the target network.
```

Author: Done.

> Man-in-the-middle attacks work against any device and operating system as the attack is executed on OSI Layer 2 through ARP Spoofing. When you are MITM you might not be able to see clear text data, as the data in transit might be encrypted by using TLS, but it will give you valuable information about the hosts involved, the protocols used and the ports the app is communicating with.

#### Simulating a Man-in-the-Middle Attack with an Access Point

##### Network Setup

> Simple way to simulate a man-in-the-middle (MITM) attack is to configure network where all packets between devices and the target network are going through your machine. In mobile world, this is achieved by placing Access Point between mobile devices and your machine. Your machine is a router and an Access Point.

Reviewer 2: ```suggestion
A simple way to simulate a man-in-the-middle (MITM) attack is to configure a network where all packets between the devices in scope and the target network are going through your machine. In a mobile penetration test, this can be achieved by using an Access Point the mobile devices and your machine are connected to. Your machine is then becoming a router and an Access Point.
```

Following scenarios are possible:

> - use your machine built-in WiFi card as an Access Point and use your wired connection to connect to the target network,

Reviewer 2: Please make a newline after "possilbe:", otherwise it's a markdown linting error

> - use your machine built-in WiFi card as an Access Point and use your wired connection to connect to the target network,

Reviewer 2: ```suggestion
- Use your machines built-in WiFi card as an Access Point and use your wired connection to connect to the target network.
```

Reviewer: @sushi2k, you meant "your machine's", right?

Author: Done.

> - use external USB WiFi card as an Access Point and user your machine built-in WiFi to connect to the target network (can be vice-versa) or

Reviewer 2: ```suggestion
- Use an external USB WiFi card as an Access Point and user your machine built-in WiFi to connect to the target network (can be vice-versa).
```

> - use external Access Point and redirect traffic to your machine.

Reviewer 2: ```suggestion
- Use a separate Access Point and redirect the traffic to your machine.
```

> Scenarios with WiFi card Access Point require that WiFi card has AP (Access Point) capabilities. Additionally, you need to install tools and/or configure network to reflect man-in-the-middle position (see below).

Reviewer 2: ```suggestion
The scenario with an external USB card require that the WiFi card has the capability to create an Access Point. Additionally, you need to install some tools and/or configure the network to enforce a man-in-the-middle position (see below).
```

> Scenario with external Access Point requires access to the configuration of the AP and you should check first if the AP supports either:

Reviewer 2: ```suggestion
The scenario with a separate Access Point requires access to the configuration of the AP and you should check first if the AP supports either:
```

- port forwarding or
- has a span or mirror port.

> In both cases the AP needs to be configured to point to your machines IP. Your machine must be connected to the AP (via wired connection or Wifi) and you need to have connection to the target network (can be by the same connection as to the AP). Some additional configuration may be required on your machine to route traffic to the target network.

Reviewer 2: ```suggestion
In both cases the AP needs to be configured to point to your machines IP. Your machine must be connected to the AP (via wired connection or Wifi) and you need to have connection to the target network (can be the same connection as to the AP). Some additional configuration may be required on your machine to route traffic to the target network.
```

Reviewer: @sushi2k could you please enhance your suggestion with..?

Your machine's IP
WiFi

> > If external Access Point belongs to customer, all changes and configurations should be clarified prior to the engagement.

Reviewer 2: ```suggestion
> If the external Access Point belongs to the customer, all changes and configurations should be clarified prior to the engagement and a backup should be created, before making any changes.
```

<img src="Images/Chapters/0x04f/architecture_MITM_AP.png" alt="Network Diagram - MITM with an Access Point">

##### Installation

> The following procedure is setting up a man-in-the-middle position using an Access Point and additional network interface:

Reviewer 2: ```suggestion
The following procedure is setting up a man-in-the-middle position using an Access Point and an additional network interface:
```

> 1.  Create an Access Point and its network.

Reviewer 2: ```suggestion
1.  Create a WiFi network either through a separate Access Point or through an external USB Card or through the build-in card of your machine.
```

Reviewer: @sushi2k  built-in* instead of build-in

Author: Resolved in new commit.

> 2.  Route traffic to additional network interface where traffic can reach the target network. Additional network interface can be wired connection or other Wifi card.

Reviewer 2: ```suggestion
2.  In case of a separate Access Point, route the traffic to your machine. In case of an external USB Card or build-in WiFi card the traffic is already available on your machine.
3. Route the incoming traffic coming from the WiFi to the additional network interface where the traffic can reach the target network. Additional network interface can be wired connection or other Wifi card, depending on your setup.
```

Reviewer: @sushi2k same here built-in

Reviewer: And "WiFi"

Author: Done.

> This can be done by using the built-in utilities on macOS. You can use [share the internet connection on Mac with other network users](https://support.apple.com/en-ke/guide/mac-help/mchlp1540/mac "Share the internet connection on Mac with other network users").

Reviewer 2: This should be part of the first step, creating a Wifi.

> For all major Linux and Unix operating systems you need:

Reviewer: Please remove the "or" from the 2nd last bullet point. Remove the last bullet point and the final commas. This is according to our style guide. Thank you! :)

```suggestion
For all major Linux and Unix operating systems you need tools such as:
```

Author: Done.

- hostapd,
- dnsmasq,
- iptables,
- wpa_supplicant,
- airmon-ng or
- its equivalents.

For Kali Linux you can install these tools with `apt-get`:

```shell
$ apt-get update
$ apt-get install hostapd dnsmasq aircrack-ng
```
> iptables and wpa_supplicant are installed by default on Kali Linux.

> You should check if your one of your Wifi card has AP capabilities. This can be done by using the command `iwconfig` on Kali Linux:

Reviewer 2: This can be part of line 118, "Use your machines built-in WiFi card as an Access Point and use your wired connection to connect to the target network. You can verify if your Wifi card has AP capabilites..."

> You should check if your one of your Wifi card has AP capabilities. This can be done by using the command `iwconfig` on Kali Linux:

Reviewer: I agree with @sushi2k and also suggest this correction
```suggestion
You should check if your one of your WiFi card has AP capabilities. This can be done by using the command `iwconfig` on Kali Linux:
```

Author: Done.

    ```shell
    $ iw list | grep AP 
    ```

> If your Wifi card has AP capabilities, you can use it to create Access Point.

Reviewer: Please ensure that all words "WiFi" are consistent. Also for "access point" which should be lowercase
```suggestion
If your WiFi card has AP capabilities, you can use it to create an access point.
```

Author: New commit. Comment implemented.

##### Configuration

> We focus on configuration files for Kali Linux. Following values were defined:

Reviewer 2: ```suggestion
We focus on the configuration files for Kali Linux. Following values need to be defined:
```

- wlan1 - id of the AP network interface (with AP capabilities),
- wlan0 - id of the target network interface (this can be wired interface or other Wifi card)
- 10.0.0.0/24 - IP addresses and mask of AP network

> You need following configuration files:

Reviewer 2: ```suggestion
The following configuration files need to be changed and adjusted accordingly:
```

- hostapd.conf

    ```
    # Name of the WiFi interface we use 
    interface=wlan1
    # Use the nl80211 driver
    driver=nl80211
    hw_mode=g
    channel=6
    wmm_enabled=1
    macaddr_acl=0
    auth_algs=1
    ignore_broadcast_ssid=0
    wpa=2
    wpa_key_mgmt=WPA-PSK
    rsn_pairwise=CCMP
    # Name of the AP network
    ssid=STM-AP
    # Password of the AP network
    wpa_passphrase=password
    ```

- wpa_supplicant.conf

    ```
    network={
        ssid="NAME_OF_THE_TARGET_NETWORK"
        psk="PASSWORD_OF_THE_TARGET_NETWORK"
    }
    ```

- dnsmasq.conf

    ```
    interface=wlan1
    dhcp-range=10.0.0.10,10.0.0.250,12h
    dhcp-option=3,10.0.0.1
    dhcp-option=6,10.0.0.1
    server=8.8.8.8
    log-queries
    log-dhcp
    listen-address=127.0.0.1
    ```

> > These are example configs. Some adjustments may be required.

Reviewer 2: you can remove this, as it's included above now

Author: Done.

##### MITM Attack

> To be able to get a man-in-the-middle position you need to run above configuration. This can be done by using following commands on Kali Linux:

Reviewer 2: ```suggestion
To be able to get a man-in-the-middle position you need to run the above configuration. This can be done by using the following commands on Kali Linux:
```

    ```shell
    # check if other process is not using Wifi interfaces
    $ airmon-ng check kill
    # configure IP address of the AP network interface
    $ ifconfig wlan1 10.0.0.1 up
    # start access point
    $ hostapd hostapd.conf
    # connect the target network interface
    $ wpa_supplicant -B -i wlan0 -c wpa_supplicant.conf
    # run DNS server
    $ dnsmasq -C dnsmasq.conf -d
    # enable routing 
    $ echo 1 > /proc/sys/net/ipv4/ip_forward
    # iptables will NAT connections from AP network interface to the target network interface
    $ iptables --flush
    $ iptables --table nat --append POSTROUTING --out-interface wlan0 -j MASQUERADE
    $ iptables --append FORWARD --in-interface wlan1 -j ACCEPT 
    $ iptables -t nat -A POSTROUTING -j MASQUERADE
    ```

> Now you can connect mobile devices to the Access Point.

Reviewer 2: ```suggestion
Now you can connect your mobile devices to the Access Point.
```

> ##### Network Analyzer Tool

Reviewer 2: Make this #### instead of #####, then it's a separate paragraph, applicable to both scnearios above.

Author: Done.

Install a tool that allows you to monitor and analyze the network traffic that will be redirected to your machine. The two most common network monitoring (or capturing) tools are:

> - [Wireshark](https://www.wireshark.org "Wireshark") (CLI pendant: [tshark](https://www.wireshark.org/docs/man-pages/tshark.html "TShark")) and

Reviewer: ```suggestion
- [Wireshark](https://www.wireshark.org "Wireshark") (CLI pendant: [TShark](https://www.wireshark.org/docs/man-pages/tshark.html "TShark"))
```
